python/ray/serve/pipeline/json_serde.py

```diff
@@ -0,0 +1,223 @@
+    with potentially deeply nested structure with other DAGNode instances.
+
+    Enforcements:
+        - All args, kwargs and other_args_to_resolve used in Ray DAG needs to
+            be JSON serializable in order to be converted and deployed using
+            Ray Serve.
+        - All modules such as class or functions need to be visible and
+            importable on top of its file, and can be resolved via a fully
+            qualified import_path.
+        - No DAGNode instance should appear in bound .options(), which should be
+            JSON serializable with default encoder.
+    """
+
+    def default(self, obj):
+        # TODO: (jiaodong) Have better abtraction for users to extend and use
+        # JSON serialization of their own object types
+        # For replaced deployment handles used as init args or kwargs.
+        if isinstance(obj, RayServeHandle):
+            return json.dumps(obj, cls=ServeHandleEncoder)
+        # For all other DAGNode types.
+        if isinstance(obj, DAGNode):
+            dag_node = obj
+            args = dag_node.get_args()
+            kwargs = dag_node.get_kwargs()
+            options = dag_node.get_options()
+            other_args_to_resolve = dag_node.get_other_args_to_resolve()
+            # stable_uuid will be re-generated upon new constructor execution
+            # except for ClassNode used as parent of ClassMethodNode
+            result_dict = {
+                DAGNODE_TYPE_KEY: dag_node.__class__.__name__,
+                # Placeholder for better ordering
+                "import_path": "",
+                "args": json.dumps(args, cls=DAGNodeEncoder),
+                "kwargs": json.dumps(kwargs, cls=DAGNodeEncoder),
+                # .options() should not contain any DAGNode type
+                "options": json.dumps(options),
+                "other_args_to_resolve": json.dumps(
+                    other_args_to_resolve, cls=DAGNodeEncoder
+                ),
+            }
+            # TODO: (jiaodong) Support arbitrary InputNode args and pydantic
+            # input schema.
+            import_path = ""
+            if isinstance(dag_node, InputNode):
+                return result_dict
+            # TODO: (jiaodong) Handle __main__ swap to filename
+            if isinstance(dag_node, ClassNode):
+                body = dag_node._body.__ray_actor_class__
+                import_path = f"{body.__module__}.{body.__qualname__}"
+            elif isinstance(dag_node, ClassMethodNode):
+                parent_class_node = dag_node._parent_class_node
+                body = dag_node._parent_class_node._body.__ray_actor_class__
+                import_path = f"{body.__module__}.{body.__qualname__}"
+                result_dict.update(
+                    {
+                        "method_name": dag_node._method_name,
+                        "parent_class_node_stable_uuid": parent_class_node._stable_uuid,
+                    }
+                )
```

Comment:
the to_json / from_json I suggested would also make this cleaner -- each node type has its own impl